[wes-w-z-h] iP

wes-w-z-h opened the pull request:
# Mizzz

>  “Nothing is less productive than to make more efficient what should not be done at all.” – Peter Drucker [(source)](https://dansilvestre.com/productivity-quotes)

Mizz free's your mind from having to remember things you need to do. It's,

- text-based
- easy to learn
- ~FAST~ SUPER FAST to use

All you need to do is,

1.  Download it from [here.](https://github.com/wes-w-z-h/ip/releases)
2. double click it.
3. add your tasks.
4. let it manage your tasks for you 😀

AND ITS **FREE!!!**

Features:

- [X] Managing tasks
- [ ] Managing deadlines
- [ ] Reminders (coming soon)

---

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

wes-w-z-h committed:
Rename main class

wes-w-z-h committed:
Add greeting and exit

wes-w-z-h committed:
Add suppport for echo

wes-w-z-h committed:
Refactor Mizz class

wes-w-z-h committed:
Add support for store cmds

wes-w-z-h committed:
Add support for add command

wes-w-z-h committed:
Add utility class

wes-w-z-h committed:
Fix bug in list

wes-w-z-h committed:
Add Task class

wes-w-z-h committed:
Add support for marking and unmarking tasks

wes-w-z-h committed:
Add support for todo task

wes-w-z-h committed:
Rename to TaskHist

wes-w-z-h committed:
Add support for invalid idxes

wes-w-z-h committed:
Fix invalid idx check

wes-w-z-h committed:
Fix indentation for printing

wes-w-z-h committed:
Add support for deadlines and events

wes-w-z-h committed:
Add support for multiple input test files

wes-w-z-h committed:
Add test cases

wes-w-z-h committed:
Move prettyPrint to Utility

wes-w-z-h committed:
Add error handling

wes-w-z-h committed:
Move verify function to new class

wes-w-z-h committed:
Add support for delete

wes-w-z-h committed:
Update mark exception

- Update constructor to include mark or unmark
- Update toString to only return relevant command

wes-w-z-h committed:
Update name to Mizz

wes-w-z-h committed:
Rename to taskList

wes-w-z-h committed:
Add support for saving

- Add class parser to parse the input file or input string
- Add storage class to provide read and write function

wes-w-z-h committed:
Merge pull request #1 from wes-w-z-h/branch-Level-7

Add support for saving

wes-w-z-h committed:
Update parser

wes-w-z-h committed:
Merge pull request #2 from wes-w-z-h/branch-Level-7

Update parser

wes-w-z-h committed:
Refactor taskList

wes-w-z-h committed:
Add support for dates

wes-w-z-h committed:
Add custom error for invalid date

wes-w-z-h committed:
Merge pull request #3 from wes-w-z-h/branch-Level-8

Add custom error for invalid date

wes-w-z-h committed:
Fix parse format for date

wes-w-z-h committed:
Merge pull request #4 from wes-w-z-h/branch-Level-8

Fix parse format for date

wes-w-z-h committed:
Update \n to use sys.linSeparator

wes-w-z-h committed:
Add store to .gitignore

wes-w-z-h committed:
Add command class and subclasses

wes-w-z-h committed:
Add support for more commands

wes-w-z-h committed:
Merge pull request #5 from wes-w-z-h/A-MoreOOP

A more oop

wes-w-z-h committed:
Remove unused function

wes-w-z-h committed:
Fix not  being added to storage

wes-w-z-h committed:
Add gradle support (#6)

* Add Gradle support

* Bump gradle and lib version

* build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

---------

Co-authored-by: damithc <damithch@damithch-mbp.comp.nus.edu.sg>
Co-authored-by: lzq <li.zhaoqi@dhs.sg>
Co-authored-by: aureliony <39163684+aureliony@users.noreply.github.com>

wes-w-z-h committed:
Update mainClass name

wes-w-z-h committed:
Fix invalid format in file handling

wes-w-z-h committed:
Add tests for Parser

wes-w-z-h committed:
Add test for Storage class

wes-w-z-h committed:
Merge pull request #7 from wes-w-z-h/A-JUnit

A j unit

wes-w-z-h committed:
Update archiveBaseName

wes-w-z-h committed:
Add javadocs

wes-w-z-h committed:
Fix indentation

wes-w-z-h committed:
Update to follow Oracle style guide

wes-w-z-h committed:
Merge pull request #9 from wes-w-z-h/branch-A-JavaDoc

Add javadocs

wes-w-z-h committed:
Resolve merge conflict by incorporating both suggestions

wes-w-z-h committed:
Merge pull request #8 from wes-w-z-h/branch-A-CodingStandard

Branch a coding standard

wes-w-z-h committed:
Fix xml

wes-w-z-h committed:
Fix style errors

wes-w-z-h committed:
Merge pull request #10 from wes-w-z-h/branch-A-CodingStandard

Fix style errors

wes-w-z-h committed:
Add support for find

wes-w-z-h committed:
Merge pull request #11 from wes-w-z-h/branch-Level-9

Add support for find

wes-w-z-h committed:
Update gradle plugins

wes-w-z-h committed:
Update javadocs

- Add javadocs
- Fix coding standards

wes-w-z-h committed:
Merge pull request #12 from wes-w-z-h/branch-A-JavaDoc

Update javadocs

wes-w-z-h committed:
Move into package

wes-w-z-h committed:
Initial support for gui

wes-w-z-h committed:
Update support for gui

wes-w-z-h committed:
Merge pull request #13 from wes-w-z-h/branch-Level-10

Add gui

bbbbcccd commented on src/main/java/util/Utility.java:
I like how you created a Utility file for printing outputs 👍

bbbbcccd commented on src/main/java/util/Validator.java:
```diff
@@ -0,0 +1,124 @@
+     */
+    public static void verifyTodo(String[] details) throws ToDoException {
+        if (details.length <= 1) {
+            throw new ToDoException("Missing a descrption for this todo!");
```
Typo here

ty-3401 replied:
Tiny misspelling in description.

bbbbcccd reviewed (commented):
Lgtm! You're the best 🥇

cshao02 commented on src/main/java/commands/Command.java:
```diff
@@ -0,0 +1,114 @@
+     * Enum represnting all valid commands. <br>
+     * Contains useful methods to interact with valid types.
+     */
+    public static enum CommandTypes {
+        TODO, DEADLINE, EVENT, LIST, MARK, UNMARK, DELETE, BYE, FIND;
```
I like how you incorporated enums in your code here to represent the different commands

cshao02 commented on src/main/java/commands/Command.java:
```diff
@@ -0,0 +1,114 @@
+         * @throws MizzException if an invalid type is used.
+         */
+        public static CommandTypes toCommandType(String type) throws MizzException {
+            switch (type.strip().toLowerCase()) {
```
I like that you accounted for user inputs that may be capitalised

cshao02 commented on src/main/java/exceptions/DeadlineException.java:
```diff
@@ -0,0 +1,19 @@
+/**
+ * Class for a deadline exception.
+ */
+public class DeadlineException extends MizzException {
```
I like how you created Exception subclasses for each of the possible exception cases

cshao02 commented on src/main/java/tasks/Deadline.java:
```diff
@@ -0,0 +1,28 @@
+ * Class encapsulating the deadline task.
+ */
+public class Deadline extends Task {
+    private LocalDate by;
```
Perhaps a more descriptive name can be used here

cshao02 commented on src/main/java/util/Ui.java:
```diff
@@ -0,0 +1,82 @@
+     * @param strArr The array of strings that make up the msg, each element will be separated with
+     *        a \n in the printed msg.
+     */
+    public void printResponse(String... strArr) {
+        StringBuilder sb = new StringBuilder();
+
+        for (int i = 0; i < strArr.length - 1; i++) {
+            sb.append(Utility.INDENT + strArr[i] + Utility.NEW_LINE);
+        }
+        if (strArr.length >= 1) {
+            sb.append(Utility.INDENT + strArr[strArr.length - 1]);
+        }
+        Utility.prettyPrint(sb.toString());
+    }
+
```
Nice! I like how you used stringbuilder instead of string concatenation

cshao02 commented on src/main/java/util/Utility.java:
```diff
@@ -0,0 +1,22 @@
+ * Class storing constants and utility methods.
+ */
+public class Utility {
+    public static final String LINE = "-------------";
+    public static final String INDENT = "    ";
+    public static final String INDENTED_LINE = Utility.INDENT + Utility.LINE;
+    public static final String NEW_LINE = System.lineSeparator();
```
I like how you declared constants as final and used all caps for the constant names

cshao02 reviewed (commented):
LGTM. Great work! Perhaps some names can be a bit more descriptive

ty-3401 commented on src/main/java/tasks/Event.java:
```diff
@@ -0,0 +1,32 @@
+    /**
+     * Constructor for an Event
+     *
+     * @param description
+     * @param from
+     * @param to
```
I believe the @params can be omitted since they are obvious in this context.
https://se-education.org/guides/conventions/java/intermediate.html#comments

ty-3401 commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    /**
+     * Methd to mark the task as undone.
```
Just a tiny misspelling of method.

ty-3401 commented on src/main/java/util/Parser.java:
```diff
@@ -0,0 +1,212 @@
+            }
+            Task t;
+            switch (taskType) {
+            case 'T': {
```
Is it possible to change the letters to the full task name (e.g. "T" to "ToDo"). I believe it'll make the code clearer and possibly allow for scaling in the future should another task starting with "T" be needed.

ty-3401 commented on src/main/java/util/Parser.java:
```diff
@@ -0,0 +1,212 @@
+     *
+     * @param inputString The input from the scanner.
+     * @return The parsed string as an array of size 4 where:
+     *         <ul>
+     *         <li>parsedString[0] -> command</li>
+     *         <li>parsedString[1] -> description</li>
+     *         <li>parsedString[2] -> "" if todo | by if deadline | from if event</li>
+     *         <li>parsedString[3] -> to if event else ""</li>
+     *         </ul>
```
I like how you explicitly show what element in the array means what.

ty-3401 commented on src/main/java/util/Validator.java:
```diff
@@ -0,0 +1,124 @@
+            throw new UpdateMarkedException("Too few arguments missing idx to mark/ unmark",
+                    details[0]);
+        }
+        try {
+            Integer.parseInt(details[1]);
+        } catch (NumberFormatException e) {
+            throw new UpdateMarkedException(
+                    String.format("Last I checked (%s)'s no int :/", details[1]), details[0]);
+        }
+    }
+
+    /**
+     * Method to validate the delete command.
+     *
+     * @param details The split input from the user.
+     * @throws DeleteException if missing details.
+     */
+    public static void verifyDelete(String[] details) throws DeleteException {
+        if (details.length == 1) {
+            throw new DeleteException("Too few arguments! Missing idx.");
+        }
+        try {
+            Integer.parseInt(details[1]);
+        } catch (NumberFormatException e) {
+            throw new DeleteException(String.format("Last I checked (%s)'s no int :/", details[1]));
+        }
+    }
+
+    /**
+     * Method to validate the find command.
+     *
+     * @param details The split input from the user.
+     * @throws FindException if missing details.
+     */
+    public static void verifyFind(String[] details) throws FindException {
+        if (details.length <= 1) {
+            throw new FindException("Incomplete request! Too few argument, missing: keyword");
+        }
+    }
+}
```
I love the detailed error messages! Makes it much easier for the user to identify what's wrong.

lm-44 commented on src/test/java/util/ParserTest.java:
```diff
@@ -0,0 +1,91 @@
+    Path tempDir;
+
+    @Test
+    public void testParseFromStorage_withValidData() {
```
I believe all the methods here should be in camelCase. Let me know if I'm mistaken on this, since I don't believe the coding standard explicitly mentions the standard for test cases.
https://se-education.org/guides/conventions/java/intermediate.html#naming

ty-3401 reviewed (commented):
Sorry for the late review! I've mostly identified some nits that may be hard for checkstyle to pick up. Hopefully that will make your code look even more polished!

wes-w-z-h committed:
Add deps for javafx

wes-w-z-h committed:
Add assertions

* assert assumptions that will break the code
* check null or invalid indexes in the params

Assertions were chosen since these issues should not arise
from normal execution of the code unless a programmer changes
the structure/ flow of the program.

Assertions are preffered over exceptions since it is unlikely
for those conditions to happen under normal operation.

wes-w-z-h committed:
Merge pull request #14 from wes-w-z-h/branch-A-Assertions

Add assertions

wes-w-z-h committed:
Fix assertion

wes-w-z-h committed:
Update names

- plural tense for arrays
- renamed to more descriptive names

By doing so, the code quality will increase.

wes-w-z-h committed:
Add constants

- add name for the 'magic numbers' in the parsed input

The use of integers make the code harder to understand.

Naming the constants for the specific index allows readers
to more easily understand what the code is doing at each line.

wes-w-z-h committed:
Fix type

wes-w-z-h committed:
Merge pull request #15 from wes-w-z-h/branch-A-CodeQuality

Branch a code quality

wes-w-z-h committed:
Create gradle.yml

wes-w-z-h committed:
Update gradle.yml

wes-w-z-h committed:
Update gradle.yml

wes-w-z-h committed:
Merge pull request #16 from wes-w-z-h/branch-CI

Add CI

wes-w-z-h committed:
Update bulild.gradle

wes-w-z-h committed:
Add support for MassOps: delete

- Add ability to delete multiple tasks at once
- Checks all indexes that are queued for delete

Deleting tasks one by one is tedious.

Mass ops for delete allows user to delete miltiple tasks
in one command, streamlining the process.

wes-w-z-h committed:
Fix coding standard

wes-w-z-h committed:
Remove print statements

wes-w-z-h committed:
Add massOps: delete

wes-w-z-h committed:
Merge pull request #17 from wes-w-z-h/branch-C-MassOps

Add mass ops

wes-w-z-h committed:
Remove comment

wes-w-z-h committed:
Refactor creating task

wes-w-z-h committed:
Fix code quality/standard issues

wes-w-z-h committed:
Merge pull request #19 from wes-w-z-h/branch-CodeQuality-CodingStandard-fixes

Fix code quality/standard issues

wes-w-z-h committed:
Add css

wes-w-z-h committed:
Merge pull request #20 from wes-w-z-h/branch-A-BetterGui

Add css

wes-w-z-h committed:
Fix bug for duplicate idx in massOps

- Use streams to further clean input
- Only perform op on distinct idxs

wes-w-z-h committed:
Add custom font

wes-w-z-h committed:
Remove use of null

wes-w-z-h committed:
Update GUI

wes-w-z-h committed:
Update images

wes-w-z-h committed:
Add docs

wes-w-z-h committed:
Fix typo

wes-w-z-h committed:
Update image loc

wes-w-z-h committed:
Fix code quality
